logstash-core/lib/logstash/pipeline.rb

```diff
@@ -253,9 +253,16 @@ def stopped?
   def start_workers
     @worker_threads.clear # In case we're restarting the pipeline
     begin
-      start_inputs
-      @outputs.each {|o| o.register }
-      @filters.each {|f| f.register }
+      begin
+        registered = []
+        @outputs.each { |p| p.register; registered << p }
+        @filters.each { |p| p.register; registered << p }
+      rescue => e
+        @logger.error("Error registering plugin", :error => e)
+        # close all sucessfully registered plugins
+        registered.each { |p| p.do_close }
```

Comment:
excellent question. I think it would be the responsibility of the failing method to cleanup after itself because it is the only one that knows in which state it is and what needs to be cleaned. so the contact then becomes:

1- the `register` method  never creates any unknown state, it cleans state upon exception
2- only if register has correctly returned can `do_close` be safely called.